{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b3946e2b-1325-4a1b-94a8-7d3202f2f026",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os, vcf, sys, itertools, shutil, subprocess, glob, yaml\n",
    "from Bio import Entrez, Seq, SeqIO\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "H37Rv_regions = pd.read_csv(\"./references/ref_genome/mycobrowser_h37rv_v4.csv\")\n",
    "isolate_metadata = pd.read_csv(\"/n/data1/hms/dbmi/farhat/rollingDB/metadata/isolate_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2b270213-7f4e-49ee-bd26-e13e68005511",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "354"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kraken_files = glob.glob(\"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/*/*/kraken/kraken_report\")\n",
    "len(kraken_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "572e0563-9ac5-4055-a33e-3e2ffd369e04",
   "metadata": {},
   "outputs": [],
   "source": [
    "high_unclassified = []\n",
    "\n",
    "for fName in kraken_files:\n",
    "    df = pd.read_csv(fName, sep='\\t', header=None)\n",
    "\n",
    "    if df.loc[df[5]=='unclassified'][0].values[0] > 25:\n",
    "        high_unclassified.append(fName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c6d2f80d-ab6a-4b85-9d87-25c21c44bb1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.listdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4c45ded0-fff5-4eab-801d-337f4cde864c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-179/MFS-179/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-183/MFS-183/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-304/MFS-304/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-422/MFS-422/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-423/MFS-423/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-425/MFS-425/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-451/MFS-451/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-453/MFS-453/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-477/MFS-477/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-490/MFS-490/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-491/MFS-491/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-492/MFS-492/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-493/MFS-493/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-494/MFS-494/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-495/MFS-495/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-496/MFS-496/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-497/MFS-497/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-498/MFS-498/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-499/MFS-499/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-500/MFS-500/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-501/MFS-501/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-502/MFS-502/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-503/MFS-503/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-504/MFS-504/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-505/MFS-505/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-506/MFS-506/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-507/MFS-507/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-508/MFS-508/kraken/kraken_report',\n",
       " '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-78/MFS-78/kraken/kraken_report']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "high_unclassified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0aa747d7-3bb3-47d5-8f07-c5847365645f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for thing in high_unclassified:\n",
    "    sample_id = os.path.basename(os.path.dirname(os.path.dirname(thing)))\n",
    "\n",
    "    for dir_name in [f'{sample_id}/bam', 'bam', 'pilon', 'WHO_resistance', f'{sample_id}/fastlin', 'lineage']:\n",
    "        full_dir = f\"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/{sample_id}/{dir_name}\"\n",
    "        if os.path.isdir(full_dir):\n",
    "            shutil.rmtree(full_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a482a119-62b1-4079-89b6-a026f9f53b79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-179/MFS-179/bam\n",
      "/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-179/bam\n"
     ]
    }
   ],
   "source": [
    "name = 'MFS-179'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea4f385e-03b1-4d30-bc7e-9fa9f76d3bab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3be06330-3398-47d8-800b-636ca86fa97a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MFS-179.dedup.bam', 'MFS-179.dedup.bam.bai', 'MFS-179.dedup.bam.metrics']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(\"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-179/MFS-179/bam\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "98ce1cfd-fb66-4e59-93d8-6935be585fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-605/MFS-605/kraken/kraken_report', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "d8d5d4a2-7e12-4901-a9ce-78d459311a39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18.03</td>\n",
       "      <td>1230659</td>\n",
       "      <td>1230659</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>R1</td>\n",
       "      <td>131567</td>\n",
       "      <td>cellular organisms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>D</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>D1</td>\n",
       "      <td>1783272</td>\n",
       "      <td>Terrabacteria group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>P</td>\n",
       "      <td>201174</td>\n",
       "      <td>Actinomycetota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>1760</td>\n",
       "      <td>Actinomycetes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>O</td>\n",
       "      <td>85007</td>\n",
       "      <td>Mycobacteriales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>F</td>\n",
       "      <td>1762</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>1763</td>\n",
       "      <td>Mycobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>0</td>\n",
       "      <td>G1</td>\n",
       "      <td>77643</td>\n",
       "      <td>Mycobacterium tuberculosis c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>81.97</td>\n",
       "      <td>5594310</td>\n",
       "      <td>5588774</td>\n",
       "      <td>S</td>\n",
       "      <td>1773</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.04</td>\n",
       "      <td>3013</td>\n",
       "      <td>3013</td>\n",
       "      <td>S1</td>\n",
       "      <td>83331</td>\n",
       "      <td>Mycobacterium tuberculos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.01</td>\n",
       "      <td>889</td>\n",
       "      <td>889</td>\n",
       "      <td>S1</td>\n",
       "      <td>707235</td>\n",
       "      <td>Mycobacterium tuberculos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.01</td>\n",
       "      <td>876</td>\n",
       "      <td>876</td>\n",
       "      <td>S1</td>\n",
       "      <td>336982</td>\n",
       "      <td>Mycobacterium tuberculos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.01</td>\n",
       "      <td>758</td>\n",
       "      <td>758</td>\n",
       "      <td>S1</td>\n",
       "      <td>83332</td>\n",
       "      <td>Mycobacterium tuberculos...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0        1        2   3        4  \\\n",
       "0   18.03  1230659  1230659   U        0   \n",
       "1   81.97  5594310        0   R        1   \n",
       "2   81.97  5594310        0  R1   131567   \n",
       "3   81.97  5594310        0   D        2   \n",
       "4   81.97  5594310        0  D1  1783272   \n",
       "5   81.97  5594310        0   P   201174   \n",
       "6   81.97  5594310        0   C     1760   \n",
       "7   81.97  5594310        0   O    85007   \n",
       "8   81.97  5594310        0   F     1762   \n",
       "9   81.97  5594310        0   G     1763   \n",
       "10  81.97  5594310        0  G1    77643   \n",
       "11  81.97  5594310  5588774   S     1773   \n",
       "12   0.04     3013     3013  S1    83331   \n",
       "13   0.01      889      889  S1   707235   \n",
       "14   0.01      876      876  S1   336982   \n",
       "15   0.01      758      758  S1    83332   \n",
       "\n",
       "                                                    5  \n",
       "0                                        unclassified  \n",
       "1                                                root  \n",
       "2                                  cellular organisms  \n",
       "3                                            Bacteria  \n",
       "4                                 Terrabacteria group  \n",
       "5                                      Actinomycetota  \n",
       "6                                       Actinomycetes  \n",
       "7                                     Mycobacteriales  \n",
       "8                                    Mycobacteriaceae  \n",
       "9                                       Mycobacterium  \n",
       "10                    Mycobacterium tuberculosis c...  \n",
       "11                         Mycobacterium tuberculosis  \n",
       "12                        Mycobacterium tuberculos...  \n",
       "13                        Mycobacterium tuberculos...  \n",
       "14                        Mycobacterium tuberculos...  \n",
       "15                        Mycobacterium tuberculos...  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dfe1b79-b9ea-4179-858f-ecd47a6c0035",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cc42926-ae49-41d6-b3af-da3a7e1b7fb8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "865ad626-ddd2-432a-9476-92e67f39d071",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "635"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FQ_names = glob.glob(\"/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Illumina/TRUST.SRWGS.FQs.AllSets.RenamedAndLanesMerged/*/*\")\n",
    "samples_with_FQ = np.unique([os.path.basename(fName).split('_')[0] for fName in FQ_names])\n",
    "len(samples_with_FQ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2e23aa53-2eab-44e6-b7d0-8ce4ec02b9c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame({'a': samples_with_FQ, 'b': samples_with_FQ, 'c': 0})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ab6406ec-d409-44a2-a66f-63dbe1c54f9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df['a'].isin(['MFS-53', 'MFS-270', 'MFS-280'])].to_csv(\"TRUST_samples_quick.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "884f8106-595f-4886-852e-96a66813764c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [a, b, c]\n",
       "Index: []"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query(\"a=='MFS-280'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fb378b6-75db-4d0a-bac8-a2b9c5f4ea46",
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --snakefile snakefile_2 --use-conda --conda-frontend conda --rerun-incomplete --keep-going --config isolates_to_run=\"TRUST_samples_quick.tsv\" --cores 1 --directory snakemake_small"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "11c79206-5780-4f87-a5c9-1fd56d6a3183",
   "metadata": {},
   "outputs": [],
   "source": [
    ".to_csv(\"TRUST_isolates_to_run.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1c180a4d-88cf-43c4-bac4-2ed962c6f47a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "171"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fastq_dir = '/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Illumina/TRUST.SRWGS.FQs.AllSets.RenamedAndLanesMerged/TRUST.SRWGS.FQs.Set3'\n",
    "FQ_names = glob.glob(f\"{fastq_dir}/*\")\n",
    "samples_with_FQ = np.unique([os.path.basename(fName).split('_')[0] for fName in FQ_names])\n",
    "len(samples_with_FQ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6e698c54-d66d-40a8-a828-28809d553153",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_genome = [(seq.id, seq.seq) for seq in SeqIO.parse(\"references/ref_genome/H37Rv_NC_000962.3.fna\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c5623e8c-466c-45d5-8544-af6e87e38b11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4411532"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ref_genome[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83fe0230-3f20-4703-a28f-1e399bdb37d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "4411532*2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf1ca1e-b7fb-49dc-a3bd-16e1a2c96364",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fastq_dir = '/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Illumina/TRUST.SRWGS.FQs.AllSets.RenamedAndLanesMerged/TRUST.SRWGS.FQs.Set3'\n",
    "# FQ_names = glob.glob(f\"{fastq_dir}/*\")\n",
    "# samples_with_FQ = np.unique([os.path.basename(fName).split('_')[0] for fName in FQ_names])\n",
    "# print(len(samples_with_FQ))\n",
    "\n",
    "# for i, name in enumerate(samples_with_FQ):\n",
    "    \n",
    "#     new_dir = f\"/n/scratch/users/s/sak0914/TRUST_FASTQ/{name}\"\n",
    "    \n",
    "#     if not os.path.isdir(new_dir):\n",
    "#         os.mkdir(new_dir)\n",
    "\n",
    "#     if not os.path.isfile(f\"{new_dir}/{name}_R1.fastq.gz\"):\n",
    "#         shutil.copy(f\"{fastq_dir}/{name}_R1.fastq.gz\", f\"{new_dir}/{name}_R1.fastq.gz\")\n",
    "\n",
    "#     if not os.path.isfile(f\"{new_dir}/{name}_R2.fastq.gz\"):\n",
    "#         shutil.copy(f\"{fastq_dir}/{name}_R2.fastq.gz\", f\"{new_dir}/{name}_R2.fastq.gz\")\n",
    "\n",
    "#     if i % 50 == 0:\n",
    "#         print(i)\n",
    "\n",
    "# fastq_dir = '/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Illumina/TRUST.SRWGS.FQs.AllSets.RenamedAndLanesMerged/TRUST.SRWGS.FQs.Set2'\n",
    "# FQ_names = glob.glob(f\"{fastq_dir}/*\")\n",
    "# samples_with_FQ = np.unique([os.path.basename(fName).split('_')[0] for fName in FQ_names])\n",
    "# print(len(samples_with_FQ))\n",
    "\n",
    "# for i, name in enumerate(samples_with_FQ):\n",
    "    \n",
    "#     new_dir = f\"/n/scratch/users/s/sak0914/TRUST_FASTQ/{name}\"\n",
    "    \n",
    "#     if not os.path.isdir(new_dir):\n",
    "#         os.mkdir(new_dir)\n",
    "\n",
    "#     if not os.path.isfile(f\"{new_dir}/{name}_R1.fastq.gz\"):\n",
    "#         shutil.copy(f\"{fastq_dir}/{name}_R1.fastq.gz\", f\"{new_dir}/{name}_R1.fastq.gz\")\n",
    "\n",
    "#     if not os.path.isfile(f\"{new_dir}/{name}_R2.fastq.gz\"):\n",
    "#         shutil.copy(f\"{fastq_dir}/{name}_R2.fastq.gz\", f\"{new_dir}/{name}_R2.fastq.gz\")\n",
    "\n",
    "#     if i % 50 == 0:\n",
    "#         print(i)\n",
    "\n",
    "\n",
    "# fastq_dir = '/n/data1/hms/dbmi/farhat/fastq_db/TRUST_Illumina/TRUST.SRWGS.FQs.AllSets.RenamedAndLanesMerged/TRUST.SRWGS.FQs.Set4'\n",
    "# FQ_names = glob.glob(f\"{fastq_dir}/*\")\n",
    "# samples_with_FQ = np.unique([os.path.basename(fName).split('_')[0] for fName in FQ_names])\n",
    "# print(len(samples_with_FQ))\n",
    "\n",
    "# for i, name in enumerate(samples_with_FQ):\n",
    "    \n",
    "#     new_dir = f\"/n/scratch/users/s/sak0914/TRUST_FASTQ/{name}\"\n",
    "    \n",
    "#     if not os.path.isdir(new_dir):\n",
    "#         os.mkdir(new_dir)\n",
    "\n",
    "#     if not os.path.isfile(f\"{new_dir}/{name}_R1.fastq.gz\"):\n",
    "#         shutil.copy(f\"{fastq_dir}/{name}_R1.fastq.gz\", f\"{new_dir}/{name}_R1.fastq.gz\")\n",
    "\n",
    "#     if not os.path.isfile(f\"{new_dir}/{name}_R2.fastq.gz\"):\n",
    "#         shutil.copy(f\"{fastq_dir}/{name}_R2.fastq.gz\", f\"{new_dir}/{name}_R2.fastq.gz\")\n",
    "\n",
    "#     if i % 50 == 0:\n",
    "#         print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a36fbf93-adb0-4d3e-bbe4-d307026e944e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b738798b-0351-4160-a455-d03a795d9db3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "000bc8c2-949a-41c4-9035-17364c6878d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for dir_name in glob.glob(\"/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/*\"):\n",
    "    # shutil.rmtree(dir_name)\n",
    "    if dir_name != '/n/data1/hms/dbmi/farhat/rollingDB/TRUST/WGS/MFS-1':\n",
    "        shutil.rmtree(dir_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11ab22a-657c-432b-bb86-1f905a583295",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4960b588-6ee2-4e3c-9bca-65bcdab22917",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1a40dde-18c7-4478-928c-bbeecc3e1570",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((151978, 50), (158451, 50), 146442)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_MTBC_metadata = pd.read_csv(\"SRA_all/human_MTBC_metadata.csv.gz\", compression='gzip')\n",
    "\n",
    "human_MTBC_human_host = pd.concat([human_MTBC_metadata.dropna(subset='host').query(\"host.str.contains('|'.join(['CZ', 'human', 'homo', 'lab', 'not', 'missing', 'Unknown', 'organ', 'culture', 'humain', 'tuberculosis']), case=False) & ~host.str.contains('primate')\"),\n",
    "                     human_MTBC_metadata.loc[pd.isnull(human_MTBC_metadata['host'])]\n",
    "                    ])\n",
    "\n",
    "human_MTBC_human_host.shape, human_MTBC_metadata.shape, human_MTBC_human_host.BioSample.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3cc24b61-874a-4350-a0f0-f0cf5779745b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "145440"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "todo_samples = list(set(human_MTBC_human_host['BioSample']) - (set(os.listdir(\"/n/scratch/users/s/sak0914/smk_output\"))))\n",
    "len(todo_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3cf52bb1-7d1d-420c-ae75-d8ce75ffe136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6449"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonhuman_samples = list(set(human_MTBC_metadata.BioSample) - set(human_MTBC_human_host.BioSample))\n",
    "len(nonhuman_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c7d9d9cb-bbd3-453d-85de-5e7e420aeb26",
   "metadata": {},
   "outputs": [],
   "source": [
    "for name in nonhuman_samples:\n",
    "    if name in os.listdir(\"/n/scratch/users/s/sak0914/smk_output\"):\n",
    "        print(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f310af9-fe31-4fe5-bf5d-fca13d17e54b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "69795259-f6e1-4268-9cfb-2c1e78c6ee07",
   "metadata": {},
   "outputs": [],
   "source": [
    "isolates = pd.read_csv(\"SRA_all/isolates_to_run.tsv\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "898bc461-27ad-47b1-8ced-c72f6cbc7631",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab4a81c3-ef2f-46c0-a702-2a473c99bb26",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "af860e50-14f2-452d-85c1-6425384a008e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "949"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_finished = 0\n",
    "\n",
    "for name in isolates[0].values:\n",
    "    if os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{name}/pilon/{name}_full.vcf.gz\"):\n",
    "        count_finished += 1\n",
    "\n",
    "count_finished"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "32330e54-8daf-4435-b646-2958ac943040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "379"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(isolates[0].values).index('SAMEA10030115')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "835ee343-0dc9-4b4a-b856-6352c327dae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_search = os.listdir(\"/n/data1/hms/dbmi/farhat/rusham/Spain_Georgia_400_Isolates\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a59e9d8c-de55-4c0d-a206-2c1fabdcf931",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROLLINGDB_ID</th>\n",
       "      <th>BIOSAMPLE_ACCESSION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>RUN</th>\n",
       "      <th>DB_OF_ORIGIN</th>\n",
       "      <th>ISOLATION_DATE</th>\n",
       "      <th>ISOLATION_COUNTRY</th>\n",
       "      <th>ISOLATION_REGION</th>\n",
       "      <th>SPUTUM/CULTURE</th>\n",
       "      <th>UNPAIRED/PAIRED</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SPECIES</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Combined_Runs</th>\n",
       "      <th>Num_Runs</th>\n",
       "      <th>OtherNames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70295</th>\n",
       "      <td>SAMN09651725</td>\n",
       "      <td>SAMN09651725</td>\n",
       "      <td>SRS3542704</td>\n",
       "      <td>SRR7516399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6/30/15</td>\n",
       "      <td>Georgia</td>\n",
       "      <td>Tbilisi</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>2018-07-12 18:17:14</td>\n",
       "      <td>2018-07-12 17:42:23</td>\n",
       "      <td>SRR7516399</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ROLLINGDB_ID BIOSAMPLE_ACCESSION      SAMPLE         RUN DB_OF_ORIGIN  \\\n",
       "70295  SAMN09651725        SAMN09651725  SRS3542704  SRR7516399          NaN   \n",
       "\n",
       "      ISOLATION_DATE ISOLATION_COUNTRY ISOLATION_REGION  SPUTUM/CULTURE  \\\n",
       "70295        6/30/15           Georgia          Tbilisi               1   \n",
       "\n",
       "       UNPAIRED/PAIRED  Platform                Model  \\\n",
       "70295                1  ILLUMINA  Illumina HiSeq 2500   \n",
       "\n",
       "                          SPECIES   TaxID          ReleaseDate  \\\n",
       "70295  Mycobacterium tuberculosis  1773.0  2018-07-12 18:17:14   \n",
       "\n",
       "                  LoadDate Combined_Runs  Num_Runs OtherNames  \n",
       "70295  2018-07-12 17:42:23    SRR7516399       1.0        NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata.query(\"SAMPLE in @samples_search\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "900aa12b-aaef-4d9a-87ba-b7b3286bd00f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.Series(samples_search).to_csv(\"Rusham_samples.csv\", index=False, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "44cba0ea-663a-4922-bbf6-1384d0335d77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "benchmarks\n",
      "references\n",
      "slurm_logs\n"
     ]
    }
   ],
   "source": [
    "samples_search_2 = []\n",
    "\n",
    "for name in samples_search:\n",
    "    if not name.startswith('ERS') and not name.startswith('SRS'):\n",
    "        print(name)\n",
    "    else:\n",
    "        samples_search_2.append(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4582d922-31c7-46fc-a289-33319762afab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(399, 402)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(samples_search_2), len(samples_search)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "114024c9-0dd6-413c-a0a3-58890fc69104",
   "metadata": {},
   "outputs": [],
   "source": [
    "for name in samples_search_2:\n",
    "    shutil.rmtree(f\"/n/data1/hms/dbmi/farhat/rusham/Spain_Georgia_400_Isolates/{name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e47bf27-afd0-420a-b07e-81679ce73ee8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "dcc7f296-5d52-47e2-8a1c-69e131b794e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((151978, 50), (158451, 50))"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test.shape, human_MTBC_metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "3d4b0432-06b4-48cd-8cd1-84609ff2f750",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "82278279-0301-4235-91c3-c8d9504b005e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROLLINGDB_ID</th>\n",
       "      <th>BIOSAMPLE_ACCESSION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>RUN</th>\n",
       "      <th>DB_OF_ORIGIN</th>\n",
       "      <th>ISOLATION_DATE</th>\n",
       "      <th>ISOLATION_COUNTRY</th>\n",
       "      <th>ISOLATION_REGION</th>\n",
       "      <th>SPUTUM/CULTURE</th>\n",
       "      <th>UNPAIRED/PAIRED</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SPECIES</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Combined_Runs</th>\n",
       "      <th>Num_Runs</th>\n",
       "      <th>OtherNames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7062</th>\n",
       "      <td>YA00054057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063</th>\n",
       "      <td>YA00054080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7064</th>\n",
       "      <td>YA00059555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7065</th>\n",
       "      <td>YA00062798</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7066</th>\n",
       "      <td>YA00069278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7139</th>\n",
       "      <td>YA00173818</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7140</th>\n",
       "      <td>YA00185913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7141</th>\n",
       "      <td>YA00185914</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7142</th>\n",
       "      <td>YA00190161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7143</th>\n",
       "      <td>YA00194580</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gandhi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ROLLINGDB_ID BIOSAMPLE_ACCESSION SAMPLE  RUN DB_OF_ORIGIN ISOLATION_DATE  \\\n",
       "7062   YA00054057                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7063   YA00054080                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7064   YA00059555                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7065   YA00062798                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7066   YA00069278                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "...           ...                 ...    ...  ...          ...            ...   \n",
       "7139   YA00173818                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7140   YA00185913                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7141   YA00185914                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7142   YA00190161                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "7143   YA00194580                 NaN    NaN  NaN       Gandhi            NaN   \n",
       "\n",
       "     ISOLATION_COUNTRY ISOLATION_REGION  SPUTUM/CULTURE  UNPAIRED/PAIRED  \\\n",
       "7062               NaN              NaN               1                1   \n",
       "7063               NaN              NaN               1                1   \n",
       "7064               NaN              NaN               1                1   \n",
       "7065               NaN              NaN               1                1   \n",
       "7066               NaN              NaN               1                1   \n",
       "...                ...              ...             ...              ...   \n",
       "7139               NaN              NaN               1                1   \n",
       "7140               NaN              NaN               1                1   \n",
       "7141               NaN              NaN               1                1   \n",
       "7142               NaN              NaN               1                1   \n",
       "7143               NaN              NaN               1                1   \n",
       "\n",
       "     Platform Model                     SPECIES  TaxID ReleaseDate LoadDate  \\\n",
       "7062      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7063      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7064      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7065      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7066      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "...       ...   ...                         ...    ...         ...      ...   \n",
       "7139      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7140      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7141      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7142      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "7143      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "\n",
       "     Combined_Runs  Num_Runs OtherNames  \n",
       "7062           NaN       NaN        NaN  \n",
       "7063           NaN       NaN        NaN  \n",
       "7064           NaN       NaN        NaN  \n",
       "7065           NaN       NaN        NaN  \n",
       "7066           NaN       NaN        NaN  \n",
       "...            ...       ...        ...  \n",
       "7139           NaN       NaN        NaN  \n",
       "7140           NaN       NaN        NaN  \n",
       "7141           NaN       NaN        NaN  \n",
       "7142           NaN       NaN        NaN  \n",
       "7143           NaN       NaN        NaN  \n",
       "\n",
       "[82 rows x 19 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata.query(\"ROLLINGDB_ID.str.startswith('YA')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ec0542-14d5-435b-a592-666de50ab1e9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "5aaabd3b-c86a-4662-9730-e65d84d82239",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'African Elephant',\n",
       " 'BADGER',\n",
       " 'BOVINE',\n",
       " 'Badger',\n",
       " 'Beef',\n",
       " 'Bison',\n",
       " 'Bison bison athabascae',\n",
       " 'Bison bonasus',\n",
       " 'Bobcat',\n",
       " 'Bos indicus',\n",
       " 'Bos taurus',\n",
       " 'Bovine',\n",
       " 'CERVINE',\n",
       " 'COW',\n",
       " 'Canis',\n",
       " 'Canis lupus familiaris',\n",
       " 'Canis lupus familiaris,',\n",
       " 'Capra hircus',\n",
       " 'Capreolus capreolus',\n",
       " 'Cattle',\n",
       " 'Cervid',\n",
       " 'Cervine',\n",
       " 'Cervus elaphus',\n",
       " 'Cheese',\n",
       " 'Coyote',\n",
       " 'DEER',\n",
       " 'Dairy cattle',\n",
       " 'Dama dama',\n",
       " 'Dassie',\n",
       " 'Deer',\n",
       " 'Elephant',\n",
       " 'Elephas maximus',\n",
       " 'Elk',\n",
       " 'FERRET',\n",
       " 'Feline',\n",
       " 'Felis catus',\n",
       " 'Ferret',\n",
       " 'Fox',\n",
       " 'Goat',\n",
       " 'Hedgehog',\n",
       " 'Helarctos malayanus',\n",
       " 'Hydrochoerus hydrochaeris',\n",
       " 'Jaguar',\n",
       " 'Lama glama',\n",
       " 'Loxodonta',\n",
       " 'Loxodonta africana',\n",
       " 'Lutra lutra',\n",
       " 'Lycaon pictus',\n",
       " 'Macaca fascicularis',\n",
       " 'Meles meles',\n",
       " 'Monkey',\n",
       " 'Mouse',\n",
       " 'Mus musculus',\n",
       " 'Opossum',\n",
       " 'Oryx zoo',\n",
       " 'Otaria byronia',\n",
       " 'Ovis aries',\n",
       " 'PORCINE',\n",
       " 'POSSUM',\n",
       " 'Panthera leo',\n",
       " 'Panthera pardus',\n",
       " 'Papio',\n",
       " 'Phacochoerus africanus',\n",
       " 'Pig',\n",
       " 'Possum',\n",
       " 'Primate',\n",
       " 'Propithecus coquereli',\n",
       " 'Raccoon',\n",
       " 'Racoon',\n",
       " 'Red Deer',\n",
       " 'Red deer',\n",
       " 'Rhino',\n",
       " 'STOAT',\n",
       " 'Sheep',\n",
       " 'Stoat',\n",
       " 'Suricat',\n",
       " 'Sus scrofa',\n",
       " 'Sus scrofa domesticus',\n",
       " 'Syncerus caffer',\n",
       " 'Tissue: Lymph nodes',\n",
       " 'Tragelaphus strepsiceros',\n",
       " 'Ursus thibetanus',\n",
       " 'Varecia variegata subcincta',\n",
       " 'Vulpes vulpes',\n",
       " 'Watusi',\n",
       " 'Wild Boar',\n",
       " 'Wild boar',\n",
       " 'antelope zoo',\n",
       " 'antilope zoo',\n",
       " 'bovine',\n",
       " 'canine',\n",
       " 'cattle',\n",
       " 'cow',\n",
       " 'elk',\n",
       " 'goat',\n",
       " 'mouse',\n",
       " 'non human primate',\n",
       " 'sea lion',\n",
       " 'waterbuck zoo',\n",
       " 'wild boar'}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(human_MTBC_metadata.host.unique()) - set(df_test.host.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "61becf2a-18b4-4539-a1cf-4624278fcf32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Homo sapiens', nan, 'Elephas maximus', 'CZ1019-16', 'CZ105-19',\n",
       "       'CZ139-19', 'CZ149-18', 'CZ178-18', 'CZ188-16', 'CZ188', 'CZ19-15',\n",
       "       'CZ195', 'CZ217', 'CZ218', 'CZ221', 'CZ222', 'CZ223', 'CZ270',\n",
       "       'CZ276', 'CZ279-19', 'CZ279', 'CZ280', 'CZ282', 'CZ286', 'CZ29-19',\n",
       "       'CZ291', 'CZ292', 'CZ297', 'CZ317-15', 'CZ318-15', 'CZ318-19',\n",
       "       'CZ35-19', 'CZ354-19', 'CZ36-19', 'CZ367-17', 'CZ372-15',\n",
       "       'CZ388-19', 'CZ389-19', 'CZ395-18', 'CZ398-15', 'CZ402-19',\n",
       "       'CZ412-18', 'CZ416-15', 'CZ417-15', 'CZ421-17', 'CZ424-18',\n",
       "       'CZ445-15', 'CZ505-15', 'CZ505-18', 'CZ532-18', 'CZ541-18',\n",
       "       'CZ614-18', 'CZ628-17', 'CZ732-19', 'CZ757-18', 'CZ772-17',\n",
       "       'CZ779-16', 'CZ800-17', 'CZ851-17', 'CZ863-17', 'CZ8842', 'CZ8946',\n",
       "       'CZ895-16', 'CZ9206', 'CZ9320', 'CZ95-17', 'CZ553-17', 'Goat',\n",
       "       'Homo sapiens sapiens', 'Mycobacterium tuberculosis', 'Bos taurus',\n",
       "       'Syncerus caffer', 'Meles meles', 'Cervus elaphus', 'Lutra lutra',\n",
       "       'Ovis aries', 'Otaria byronia', 'Panthera pardus', 'Papio',\n",
       "       'Sus scrofa domesticus', 'Sus scrofa', 'Homo_sapiens',\n",
       "       'Macaca fascicularis', 'Loxodonta africana', 'Culture', 'cattle',\n",
       "       'Phacochoerus africanus', 'Bovine', 'Lycaon pictus',\n",
       "       'Panthera leo', 'Tragelaphus strepsiceros', 'Vulpes vulpes',\n",
       "       'antelope zoo', 'Oryx zoo', 'antilope zoo', 'waterbuck zoo',\n",
       "       'Tuberculosis', 'cow', 'Cervid', 'Cattle', 'Bobcat', 'Jaguar',\n",
       "       'Primate', 'Cheese', 'Raccoon', 'Opossum', 'Racoon', 'Beef',\n",
       "       'Felis catus', 'Bison', 'missing', 'Deer', 'Elephant', 'Rhino',\n",
       "       'non human primate', 'not collected', 'Unknown', 'human patient',\n",
       "       'Dassie', 'Suricat', 'Coyote', 'wild boar', 'BOVINE', 'POSSUM',\n",
       "       'CERVINE', 'FERRET', 'PORCINE', 'STOAT', 'Homo sapiens NCBI:9606',\n",
       "       '\"Homo sapiens', 'Elk', 'Human/Culture', 'elk',\n",
       "       'Bison bison athabascae', 'African Elephant', 'Bison bonasus',\n",
       "       'Lama glama', 'sea lion', 'not applicable', 'Monkey', 'Cervine',\n",
       "       'Ferret', 'Pig', 'Possum', 'Feline', 'Hedgehog', 'Stoat',\n",
       "       'Capra hircus', 'Capreolus capreolus', 'canine', 'Laboratory',\n",
       "       'BADGER', 'human 1', 'human 2', 'human 3', 'human 4', 'human 5',\n",
       "       'human 6', 'human 7', 'human 8', 'human 9', 'human 10', 'human 11',\n",
       "       'human 12', 'human 13', 'human 14', 'human 15', 'human 16',\n",
       "       'human 17', 'human 18', 'human 19', 'human 20', 'human 21',\n",
       "       'human 22', 'human 23', 'human 24', 'human 25', 'human 26',\n",
       "       'human 27', 'human 28', 'human 29', 'human 30', 'human 31',\n",
       "       'human 32', 'human 33', 'human 34', 'human 35', 'human 36',\n",
       "       'human 37', 'human 38', 'human 39', 'human 40', 'human 41',\n",
       "       'human 42', 'human 43', 'human 44', 'human 45', 'human 46',\n",
       "       'human 47', 'human 48', 'human 49', 'human 50', 'human 51',\n",
       "       'human 52', 'human 53', 'human 54', 'human 55', 'human 56',\n",
       "       'human 57', 'human 58', 'human 59', 'human 60', 'human 61',\n",
       "       'human 62', 'human 63', 'human 64', 'human 65', 'human 66',\n",
       "       'human 67', 'human 68', 'human 69', 'human 70', 'human 71',\n",
       "       'human 72', 'human 73', 'human 74', 'human 75', 'human 76',\n",
       "       'human 77', 'human 78', 'human 79', 'human 80', 'human 81',\n",
       "       'human 82', 'human 83', 'human 84', 'human 85', 'human 86',\n",
       "       'human 87', 'human 88', 'human 89', 'human 90', 'human 91',\n",
       "       'human 92', 'human 93', 'human 94', 'human 95', 'human 96',\n",
       "       'human 97', 'human 98', 'human 99', 'human 100', 'human 101',\n",
       "       'human 102', 'human 103', 'human 104', 'human 105', 'human 106',\n",
       "       'human 107', 'human 108', 'human 109', 'human 110', 'human 111',\n",
       "       'human 112', 'human 113', 'human 114', 'human 115', 'human 116',\n",
       "       'human 117', 'human 118', 'human 119', 'human 120', 'human 121',\n",
       "       'human 122', 'human 123', 'human 124', 'human 125', 'human 126',\n",
       "       'human 127', 'human 128', 'human 129', 'human 130', 'human 131',\n",
       "       'human 132', 'human 133', 'human 134', 'human 135', 'human 136',\n",
       "       'human 137', 'human 138', 'human 139', 'human 140', 'human 141',\n",
       "       'human 142', 'human 143', 'human 144', 'human 145', 'human 146',\n",
       "       'human 147', 'human 148', 'human 149', 'human 150', 'human 151',\n",
       "       'human 152', 'human 153', 'human 154', 'Hydrochoerus hydrochaeris',\n",
       "       'COW', 'DEER', 'goat', 'Not Collected', 'Mus musculus',\n",
       "       'Body organs', 'Canis lupus familiaris,', 'Tissue: Lymph nodes',\n",
       "       'Red deer', 'Wild boar', 'Canis lupus familiaris',\n",
       "       'Not applicable', 'mouse', 'Sheep', 'Human host', 'Dairy cattle',\n",
       "       'Loxodonta', 'Watusi', 'Human, Homo sapiens', 'Canis', 'bovine',\n",
       "       'laboratory', 'Dama dama', 'Fox', 'Badger', 'Red Deer',\n",
       "       'Wild Boar', 'Lab strain', 'Varecia variegata subcincta',\n",
       "       'Propithecus coquereli', 'Helarctos malayanus', 'Ursus thibetanus',\n",
       "       'Mouse', 'Bos indicus', 'Humaine', 'Humain'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbfb6ad9-1370-4cbe-af40-483a42b808cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e66fea1e-7579-4a4a-baf8-b11c98b0b85e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e98a1e21-9d5a-423f-9c0c-be52914ce81c",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = human_MTBC_metadata[['BioSample', 'Combined_Runs']].drop_duplicates('BioSample')\n",
    "test['FASTQ'] = 1\n",
    "\n",
    "test.iloc[:1000, :].to_csv(\"SRA_all/isolates_to_run.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8375ebca-4f4b-4da5-afe3-c5630d7f5e37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA10029420</td>\n",
       "      <td>ERR7162974</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA10029421</td>\n",
       "      <td>ERR7162975</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA10029424</td>\n",
       "      <td>ERR7162976</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMEA10029425</td>\n",
       "      <td>ERR7162977</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMEA10029426</td>\n",
       "      <td>ERR7162978</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>SAMEA104027339</td>\n",
       "      <td>ERR1950004,ERR1950003</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>SAMEA104027340</td>\n",
       "      <td>ERR1950005</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>SAMEA104027341</td>\n",
       "      <td>ERR1950006</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>SAMEA104027342</td>\n",
       "      <td>ERR1950007</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>SAMEA104027343</td>\n",
       "      <td>ERR1950008,ERR1950009</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0                      1  2\n",
       "0     SAMEA10029420             ERR7162974  1\n",
       "1     SAMEA10029421             ERR7162975  1\n",
       "2     SAMEA10029424             ERR7162976  1\n",
       "3     SAMEA10029425             ERR7162977  1\n",
       "4     SAMEA10029426             ERR7162978  1\n",
       "..              ...                    ... ..\n",
       "995  SAMEA104027339  ERR1950004,ERR1950003  1\n",
       "996  SAMEA104027340             ERR1950005  1\n",
       "997  SAMEA104027341             ERR1950006  1\n",
       "998  SAMEA104027342             ERR1950007  1\n",
       "999  SAMEA104027343  ERR1950008,ERR1950009  1\n",
       "\n",
       "[1000 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"SRA_all/isolates_to_run.tsv\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "37e68253-9034-4f19-9fd4-f12b1f68dd91",
   "metadata": {},
   "outputs": [],
   "source": [
    "test.iloc[:1].to_csv(\"SRA_all/isolates_to_run.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35ac8b84-98de-48ee-9495-50f2a8f8636c",
   "metadata": {},
   "outputs": [],
   "source": [
    "/home/sak0914/Mtb_Megapipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb518de9-c77d-47a7-8d97-835d47db7812",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01cd7dce-6ffb-43fb-9577-b4a792d8e3ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4f150e5d-fd67-4322-aa39-37d7f91e7904",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_MTBC_metadata.to_csv(\"SRA_all/human_MTBC_metadata.csv.gz\", compression='gzip', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07e9bde1-c70f-4500-a26a-23dc9d67b28f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3a432998-bdb9-4a92-ad47-4817db0982a4",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "4bb75c94-2a7e-4472-bf96-9b07e2ef0f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_metadata_All_Mtb_SRA = pd.read_csv(\"../data_cleaning/isolate_metadata_All_Mtb_SRA.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "bf0ff12b-ee50-4fcc-b8cf-f4a5efb66188",
   "metadata": {},
   "outputs": [],
   "source": [
    "cass_samples_for_megapipe.columns = ['BioSample', 'Run', 'FASTQ']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "36690f7e-b2ac-43e4-a2ad-71d9b59c6184",
   "metadata": {},
   "outputs": [],
   "source": [
    "cass_samples_for_megapipe = cass_samples_for_megapipe.merge(cass_mapping_table_keep[['BioSample', 'Run', 'Name']], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "72e0241c-dcc9-4c97-8469-57673d01461b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>FASTQ</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [BioSample, Run, FASTQ, Name]\n",
       "Index: []"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_samples_for_megapipe.loc[pd.isnull(cass_samples_for_megapipe['Name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "799cdba7-dd8a-43c0-85a7-4f69a579f550",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['BioSample', 'Sample', 'Run', 'strain', 'collected_by',\n",
       "       'collection_date', 'geo_loc_name', 'host', 'host_disease',\n",
       "       'isolation_source', 'lat_lon', 'host_health_state', 'sample_name',\n",
       "       'broker name', 'ReleaseDate', 'LoadDate', 'spots', 'bases',\n",
       "       'spots_with_mates', 'avgLength', 'size_MB', 'AssemblyName',\n",
       "       'download_path', 'Experiment', 'LibraryName', 'LibraryStrategy',\n",
       "       'LibrarySelection', 'LibrarySource', 'LibraryLayout', 'InsertSize',\n",
       "       'InsertDev', 'Platform', 'Model', 'SRAStudy', 'BioProject',\n",
       "       'Study_Pubmed_id', 'ProjectID', 'SampleType', 'TaxID', 'ScientificName',\n",
       "       'SampleName', 'CenterName', 'Submission', 'RunHash', 'ReadHash',\n",
       "       'collection_date_string', 'country'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata_All_Mtb_SRA.query(\"BioSample in @cass_mapping_table_keep.BioSample\").columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "96cb55c1-d0d5-4755-b482-e575a4e613b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cass_metadata = isolate_metadata_All_Mtb_SRA.query(\"BioSample in @cass_mapping_table_keep.BioSample\")[['BioSample', 'Sample', 'Run', 'collection_date_string', 'geo_loc_name', 'country', 'ReleaseDate', 'LoadDate', 'Platform', 'Model', 'TaxID', 'ScientificName',\n",
    "       'SampleName']].sort_values(\"SampleName\").rename(columns={'geo_loc_name': 'ISOLATION_REGION', 'Sample': 'SAMPLE', 'Run': 'RUN', 'BioSample': 'BIOSAMPLE_ACCESSION', 'SampleName': 'ROLLINGDB_ID', 'ScientificName': 'SPECIES', 'country': 'ISOLATION_COUNTRY', 'collection_date_string': 'ISOLATION_DATE'})\n",
    "\n",
    "cass_metadata['Combined_Runs'] = cass_metadata['RUN']\n",
    "cass_metadata[['Num_Runs', 'SPUTUM/CULTURE', 'UNPAIRED/PAIRED']] = 1\n",
    "cass_metadata['DB_OF_ORIGIN'] = 'CASS'\n",
    "cass_metadata['OtherNames'] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "a0140eea-80f9-4f9f-9fd2-c112f73e34b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROLLINGDB_ID</th>\n",
       "      <th>BIOSAMPLE_ACCESSION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>RUN</th>\n",
       "      <th>DB_OF_ORIGIN</th>\n",
       "      <th>ISOLATION_DATE</th>\n",
       "      <th>ISOLATION_COUNTRY</th>\n",
       "      <th>ISOLATION_REGION</th>\n",
       "      <th>SPUTUM/CULTURE</th>\n",
       "      <th>UNPAIRED/PAIRED</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SPECIES</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Combined_Runs</th>\n",
       "      <th>Num_Runs</th>\n",
       "      <th>OtherNames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00-R1210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>farhat_internal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Peru</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>01-R0573_ACTTGA_L003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Murray</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>01-R0913_TAGCTT_L003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Murray</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>01-R0955_GGCTAC_L003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Murray</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>01-R1165_CGATGT_L001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Murray</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ROLLINGDB_ID BIOSAMPLE_ACCESSION SAMPLE  RUN     DB_OF_ORIGIN  \\\n",
       "0              00-R1210                 NaN    NaN  NaN  farhat_internal   \n",
       "1  01-R0573_ACTTGA_L003                 NaN    NaN  NaN           Murray   \n",
       "2  01-R0913_TAGCTT_L003                 NaN    NaN  NaN           Murray   \n",
       "3  01-R0955_GGCTAC_L003                 NaN    NaN  NaN           Murray   \n",
       "4  01-R1165_CGATGT_L001                 NaN    NaN  NaN           Murray   \n",
       "\n",
       "  ISOLATION_DATE ISOLATION_COUNTRY ISOLATION_REGION  SPUTUM/CULTURE  \\\n",
       "0            NaN              Peru             Peru               1   \n",
       "1            NaN               NaN              NaN               1   \n",
       "2            NaN               NaN              NaN               1   \n",
       "3            NaN               NaN              NaN               1   \n",
       "4            NaN               NaN              NaN               1   \n",
       "\n",
       "   UNPAIRED/PAIRED Platform Model                     SPECIES  TaxID  \\\n",
       "0                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "1                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "2                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "3                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "4                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "\n",
       "  ReleaseDate LoadDate Combined_Runs  Num_Runs OtherNames  \n",
       "0         NaN      NaN           NaN       NaN        NaN  \n",
       "1         NaN      NaN           NaN       NaN        NaN  \n",
       "2         NaN      NaN           NaN       NaN        NaN  \n",
       "3         NaN      NaN           NaN       NaN        NaN  \n",
       "4         NaN      NaN           NaN       NaN        NaN  "
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "5b2a8314-d194-4403-9b9a-e6f48534d446",
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_metadata = pd.read_csv(\"/n/data1/hms/dbmi/farhat/rollingDB/metadata/isolate_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "88243cd1-70c1-44e3-800e-f1bfdad04d6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(361, 361)"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_isolates = isolate_metadata.query(\"DB_OF_ORIGIN=='CASS'\").ROLLINGDB_ID\n",
    "len(cass_isolates), len(np.unique(cass_isolates))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "1eda194a-a05c-4ee4-b15a-5a8fd29dfef8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROLLINGDB_ID</th>\n",
       "      <th>BIOSAMPLE_ACCESSION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>RUN</th>\n",
       "      <th>DB_OF_ORIGIN</th>\n",
       "      <th>ISOLATION_DATE</th>\n",
       "      <th>ISOLATION_COUNTRY</th>\n",
       "      <th>ISOLATION_REGION</th>\n",
       "      <th>SPUTUM/CULTURE</th>\n",
       "      <th>UNPAIRED/PAIRED</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SPECIES</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Combined_Runs</th>\n",
       "      <th>Num_Runs</th>\n",
       "      <th>OtherNames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C001</td>\n",
       "      <td>SAMN13813838</td>\n",
       "      <td>SRS5977589</td>\n",
       "      <td>SRR10869296</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:34</td>\n",
       "      <td>SRR10869296</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C002</td>\n",
       "      <td>SAMN13813839</td>\n",
       "      <td>SRS5977590</td>\n",
       "      <td>SRR10869295</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:35</td>\n",
       "      <td>SRR10869295</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C003</td>\n",
       "      <td>SAMN13813716</td>\n",
       "      <td>SRS5977562</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:35</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C004</td>\n",
       "      <td>SAMN13813717</td>\n",
       "      <td>SRS5977563</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:36</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C005</td>\n",
       "      <td>SAMN13813718</td>\n",
       "      <td>SRS5977673</td>\n",
       "      <td>SRR10869211</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:33</td>\n",
       "      <td>SRR10869211</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>C479</td>\n",
       "      <td>SAMN13813835</td>\n",
       "      <td>SRS5977585</td>\n",
       "      <td>SRR10869300</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:35</td>\n",
       "      <td>SRR10869300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>C480</td>\n",
       "      <td>SAMN13814075</td>\n",
       "      <td>SRS5977852</td>\n",
       "      <td>SRR10869034</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:27</td>\n",
       "      <td>SRR10869034</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>C482</td>\n",
       "      <td>SAMN13814076</td>\n",
       "      <td>SRS5977853</td>\n",
       "      <td>SRR10869032</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:27</td>\n",
       "      <td>SRR10869032</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>C485</td>\n",
       "      <td>SAMN13813836</td>\n",
       "      <td>SRS5977587</td>\n",
       "      <td>SRR10869298</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:34</td>\n",
       "      <td>SRR10869298</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>C486</td>\n",
       "      <td>SAMN13813837</td>\n",
       "      <td>SRS5977588</td>\n",
       "      <td>SRR10869297</td>\n",
       "      <td>CASS</td>\n",
       "      <td>2019</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>1773.0</td>\n",
       "      <td>4/28/20 13:17</td>\n",
       "      <td>1/10/20 13:34</td>\n",
       "      <td>SRR10869297</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ROLLINGDB_ID BIOSAMPLE_ACCESSION      SAMPLE          RUN DB_OF_ORIGIN  \\\n",
       "0           C001        SAMN13813838  SRS5977589  SRR10869296         CASS   \n",
       "1           C002        SAMN13813839  SRS5977590  SRR10869295         CASS   \n",
       "2           C003        SAMN13813716  SRS5977562  SRR10869323         CASS   \n",
       "3           C004        SAMN13813717  SRS5977563  SRR10869322         CASS   \n",
       "4           C005        SAMN13813718  SRS5977673  SRR10869211         CASS   \n",
       "..           ...                 ...         ...          ...          ...   \n",
       "356         C479        SAMN13813835  SRS5977585  SRR10869300         CASS   \n",
       "357         C480        SAMN13814075  SRS5977852  SRR10869034         CASS   \n",
       "358         C482        SAMN13814076  SRS5977853  SRR10869032         CASS   \n",
       "359         C485        SAMN13813836  SRS5977587  SRR10869298         CASS   \n",
       "360         C486        SAMN13813837  SRS5977588  SRR10869297         CASS   \n",
       "\n",
       "    ISOLATION_DATE ISOLATION_COUNTRY ISOLATION_REGION  SPUTUM/CULTURE  \\\n",
       "0             2019      South Africa     South Africa               1   \n",
       "1             2019      South Africa     South Africa               1   \n",
       "2             2019      South Africa     South Africa               1   \n",
       "3             2019      South Africa     South Africa               1   \n",
       "4             2019      South Africa     South Africa               1   \n",
       "..             ...               ...              ...             ...   \n",
       "356           2019      South Africa     South Africa               1   \n",
       "357           2019      South Africa     South Africa               1   \n",
       "358           2019      South Africa     South Africa               1   \n",
       "359           2019      South Africa     South Africa               1   \n",
       "360           2019      South Africa     South Africa               1   \n",
       "\n",
       "     UNPAIRED/PAIRED  Platform                Model  \\\n",
       "0                  1  ILLUMINA  Illumina HiSeq 2500   \n",
       "1                  1  ILLUMINA  Illumina HiSeq 2500   \n",
       "2                  1  ILLUMINA  Illumina HiSeq 2500   \n",
       "3                  1  ILLUMINA  Illumina HiSeq 2500   \n",
       "4                  1  ILLUMINA  Illumina HiSeq 2500   \n",
       "..               ...       ...                  ...   \n",
       "356                1  ILLUMINA  Illumina HiSeq 2500   \n",
       "357                1  ILLUMINA  Illumina HiSeq 2500   \n",
       "358                1  ILLUMINA  Illumina HiSeq 2500   \n",
       "359                1  ILLUMINA  Illumina HiSeq 2500   \n",
       "360                1  ILLUMINA  Illumina HiSeq 2500   \n",
       "\n",
       "                        SPECIES   TaxID    ReleaseDate       LoadDate  \\\n",
       "0    Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:34   \n",
       "1    Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:35   \n",
       "2    Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:35   \n",
       "3    Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:36   \n",
       "4    Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:33   \n",
       "..                          ...     ...            ...            ...   \n",
       "356  Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:35   \n",
       "357  Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:27   \n",
       "358  Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:27   \n",
       "359  Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:34   \n",
       "360  Mycobacterium tuberculosis  1773.0  4/28/20 13:17  1/10/20 13:34   \n",
       "\n",
       "    Combined_Runs  Num_Runs OtherNames  \n",
       "0     SRR10869296       1.0        NaN  \n",
       "1     SRR10869295       1.0        NaN  \n",
       "2     SRR10869323       1.0        NaN  \n",
       "3     SRR10869322       1.0        NaN  \n",
       "4     SRR10869211       1.0        NaN  \n",
       "..            ...       ...        ...  \n",
       "356   SRR10869300       1.0        NaN  \n",
       "357   SRR10869034       1.0        NaN  \n",
       "358   SRR10869032       1.0        NaN  \n",
       "359   SRR10869298       1.0        NaN  \n",
       "360   SRR10869297       1.0        NaN  \n",
       "\n",
       "[361 rows x 19 columns]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata.query(\"DB_OF_ORIGIN=='CASS'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf5d4e7d-bb9e-403a-9e0a-491fe8c1163b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "125adfa3-7f2b-4992-a34b-310419b644e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cass_samples_for_megapipe.sort_values('Name')[['Name', 'Run', 'FASTQ']].to_csv(\"/home/sak0914/Mtb_Megapipe/test/CASS_isolates.tsv\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "1fcf8dfb-4b76-493b-9320-537455cea250",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Run</th>\n",
       "      <th>FASTQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>C001</td>\n",
       "      <td>SRR10869296</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>C002</td>\n",
       "      <td>SRR10869295</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>C003</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>C004</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>C005</td>\n",
       "      <td>SRR10869211</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>C479</td>\n",
       "      <td>SRR10869300</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>C480</td>\n",
       "      <td>SRR10869034</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>C482</td>\n",
       "      <td>SRR10869032</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>C485</td>\n",
       "      <td>SRR10869298</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>C486</td>\n",
       "      <td>SRR10869297</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name          Run  FASTQ\n",
       "333  C001  SRR10869296      1\n",
       "332  C002  SRR10869295      1\n",
       "360  C003  SRR10869323      1\n",
       "359  C004  SRR10869322      1\n",
       "248  C005  SRR10869211      1\n",
       "..    ...          ...    ...\n",
       "337  C479  SRR10869300      1\n",
       "71   C480  SRR10869034      1\n",
       "69   C482  SRR10869032      1\n",
       "335  C485  SRR10869298      1\n",
       "334  C486  SRR10869297      1\n",
       "\n",
       "[361 rows x 3 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_samples_for_megapipe.sort_values('Name')[['Name', 'Run', 'FASTQ']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c41b436e-714c-487b-8b6d-4d6a6c3edf3e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b2107043-15e6-4e03-a965-3d7554dbe7d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(404, 361)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_MIC_ML.query(\"DB_OF_ORIGIN=='CASS'\").ID), len(cass_mapping_table_keep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3f10f690-2eff-4fd9-bcc0-3add5ea093cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "still_missing = list(set(df_MIC_ML.query(\"DB_OF_ORIGIN=='CASS'\").ID) - set(cass_mapping_table['Sample Name']))\n",
    "len(still_missing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "92353df2-efe4-430d-a3b4-01e3925c6941",
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_metadata.loc[isolate_metadata['ROLLINGDB_ID'].isin(cass_mapping_table_keep['Name']), 'BIOSAMPLE_ACCESSION'] = isolate_metadata['ROLLINGDB_ID'].map(internalName_bioSample_mapping)\n",
    "isolate_metadata.loc[isolate_metadata['ROLLINGDB_ID'].isin(cass_mapping_table_keep['Name']), 'RUN'] = isolate_metadata['BIOSAMPLE_ACCESSION'].map(bioSample_run_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "79bdf73c-9d25-41f3-abc5-b140cf50c017",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROLLINGDB_ID</th>\n",
       "      <th>BIOSAMPLE_ACCESSION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>RUN</th>\n",
       "      <th>DB_OF_ORIGIN</th>\n",
       "      <th>ISOLATION_DATE</th>\n",
       "      <th>ISOLATION_COUNTRY</th>\n",
       "      <th>ISOLATION_REGION</th>\n",
       "      <th>SPUTUM/CULTURE</th>\n",
       "      <th>UNPAIRED/PAIRED</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SPECIES</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Combined_Runs</th>\n",
       "      <th>Num_Runs</th>\n",
       "      <th>OtherNames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>C001</td>\n",
       "      <td>SAMN13813838</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869296</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>C002</td>\n",
       "      <td>SAMN13813839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869295</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>C003</td>\n",
       "      <td>SAMN13813716</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>C004</td>\n",
       "      <td>SAMN13813717</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>C005</td>\n",
       "      <td>SAMN13813718</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869211</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1054</th>\n",
       "      <td>C479</td>\n",
       "      <td>SAMN13813835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869300</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055</th>\n",
       "      <td>C480</td>\n",
       "      <td>SAMN13814075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869034</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>C482</td>\n",
       "      <td>SAMN13814076</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869032</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>C485</td>\n",
       "      <td>SAMN13813836</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869298</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1058</th>\n",
       "      <td>C486</td>\n",
       "      <td>SAMN13813837</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR10869297</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ROLLINGDB_ID BIOSAMPLE_ACCESSION SAMPLE          RUN DB_OF_ORIGIN  \\\n",
       "698          C001        SAMN13813838    NaN  SRR10869296         CASS   \n",
       "699          C002        SAMN13813839    NaN  SRR10869295         CASS   \n",
       "700          C003        SAMN13813716    NaN  SRR10869323         CASS   \n",
       "701          C004        SAMN13813717    NaN  SRR10869322         CASS   \n",
       "702          C005        SAMN13813718    NaN  SRR10869211         CASS   \n",
       "...           ...                 ...    ...          ...          ...   \n",
       "1054         C479        SAMN13813835    NaN  SRR10869300         CASS   \n",
       "1055         C480        SAMN13814075    NaN  SRR10869034         CASS   \n",
       "1056         C482        SAMN13814076    NaN  SRR10869032         CASS   \n",
       "1057         C485        SAMN13813836    NaN  SRR10869298         CASS   \n",
       "1058         C486        SAMN13813837    NaN  SRR10869297         CASS   \n",
       "\n",
       "     ISOLATION_DATE ISOLATION_COUNTRY ISOLATION_REGION  SPUTUM/CULTURE  \\\n",
       "698             NaN               NaN              NaN               1   \n",
       "699             NaN               NaN              NaN               1   \n",
       "700             NaN               NaN              NaN               1   \n",
       "701             NaN               NaN              NaN               1   \n",
       "702             NaN               NaN              NaN               1   \n",
       "...             ...               ...              ...             ...   \n",
       "1054            NaN               NaN              NaN               1   \n",
       "1055            NaN               NaN              NaN               1   \n",
       "1056            NaN               NaN              NaN               1   \n",
       "1057            NaN               NaN              NaN               1   \n",
       "1058            NaN               NaN              NaN               1   \n",
       "\n",
       "      UNPAIRED/PAIRED Platform Model                     SPECIES  TaxID  \\\n",
       "698                 1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "699                 1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "700                 1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "701                 1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "702                 1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "...               ...      ...   ...                         ...    ...   \n",
       "1054                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "1055                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "1056                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "1057                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "1058                1      NaN   NaN  Mycobacterium tuberculosis    NaN   \n",
       "\n",
       "     ReleaseDate LoadDate Combined_Runs  Num_Runs OtherNames  \n",
       "698          NaN      NaN           NaN       NaN        NaN  \n",
       "699          NaN      NaN           NaN       NaN        NaN  \n",
       "700          NaN      NaN           NaN       NaN        NaN  \n",
       "701          NaN      NaN           NaN       NaN        NaN  \n",
       "702          NaN      NaN           NaN       NaN        NaN  \n",
       "...          ...      ...           ...       ...        ...  \n",
       "1054         NaN      NaN           NaN       NaN        NaN  \n",
       "1055         NaN      NaN           NaN       NaN        NaN  \n",
       "1056         NaN      NaN           NaN       NaN        NaN  \n",
       "1057         NaN      NaN           NaN       NaN        NaN  \n",
       "1058         NaN      NaN           NaN       NaN        NaN  \n",
       "\n",
       "[361 rows x 19 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata.loc[isolate_metadata['ROLLINGDB_ID'].isin(cass_mapping_table_keep['Name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "4e44504d-7d3e-43a7-ba69-0c60b7f49cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "internalName_bioSample_mapping = dict(zip(cass_mapping_table_keep['Name'], cass_mapping_table_keep['BioSample']))\n",
    "bioSample_run_mapping = dict(zip(cass_mapping_table_keep['BioSample'], cass_mapping_table_keep['Run']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "887fb39d-f6d2-4b7a-b22f-11d2cd2debd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "361"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(internalName_bioSample_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "f13072c3-bd92-4889-a6e5-d6a888155c18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Run                                                        SRR10868963\n",
       "Assay Type                                                         WGS\n",
       "AvgSpotLen                                                         302\n",
       "Bases                                                        392746772\n",
       "BioProject                                                 PRJNA600338\n",
       "BioSample                                                 SAMN13813815\n",
       "BioSampleModel                                             Pathogen.cl\n",
       "Bytes                                                        164614162\n",
       "Center Name                                                        UCT\n",
       "collected_by                                                      CLII\n",
       "Collection_Date                                                   2019\n",
       "Consent                                                         public\n",
       "DATASTORE filetype                                    run.zq,fastq,sra\n",
       "DATASTORE provider                                          gs,ncbi,s3\n",
       "DATASTORE region                  s3.us-east-1,ncbi.public,gs.us-east1\n",
       "Experiment                                                  SRX7539346\n",
       "geo_loc_name_country                                      South Africa\n",
       "geo_loc_name_country_continent                                  Africa\n",
       "geo_loc_name                                              South Africa\n",
       "Host_disease                                              Tuberculosis\n",
       "Host                                                      Homo sapiens\n",
       "Instrument                                         Illumina HiSeq 2500\n",
       "isolate                                                           C102\n",
       "isolation_source                                                sputum\n",
       "lat_lon                                            33.9249 S 18.4241 E\n",
       "Library Name                                                       100\n",
       "LibraryLayout                                                   PAIRED\n",
       "LibrarySelection                                                RANDOM\n",
       "LibrarySource                                                  GENOMIC\n",
       "Organism                                    Mycobacterium tuberculosis\n",
       "Platform                                                      ILLUMINA\n",
       "ReleaseDate                                       2020-04-28T00:00:00Z\n",
       "create_date                                       2020-01-10T13:24:00Z\n",
       "version                                                              1\n",
       "Sample Name                                                       C439\n",
       "SRA Study                                                    SRP241186\n",
       "strain                                                 M. tuberculosis\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table.iloc[0, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "af4a3ac0-0da0-48e2-8a10-77289cb276e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Run', 'Assay Type', 'AvgSpotLen', 'Bases', 'BioProject', 'BioSample',\n",
       "       'BioSampleModel', 'Bytes', 'Center Name', 'collected_by',\n",
       "       'Collection_Date', 'Consent', 'DATASTORE filetype',\n",
       "       'DATASTORE provider', 'DATASTORE region', 'Experiment',\n",
       "       'geo_loc_name_country', 'geo_loc_name_country_continent',\n",
       "       'geo_loc_name', 'Host_disease', 'Host', 'Instrument', 'isolate',\n",
       "       'isolation_source', 'lat_lon', 'Library Name', 'LibraryLayout',\n",
       "       'LibrarySelection', 'LibrarySource', 'Organism', 'Platform',\n",
       "       'ReleaseDate', 'create_date', 'version', 'Sample Name', 'SRA Study',\n",
       "       'strain'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2ef8d749-474c-4a26-81f6-0a8d4603b890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>Name</th>\n",
       "      <th>geo_loc_name_country</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13813815</td>\n",
       "      <td>SRR10868963</td>\n",
       "      <td>C439</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN13813814</td>\n",
       "      <td>SRR10868964</td>\n",
       "      <td>C436</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN13813813</td>\n",
       "      <td>SRR10868965</td>\n",
       "      <td>C434</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN13813812</td>\n",
       "      <td>SRR10868966</td>\n",
       "      <td>C433</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN13813811</td>\n",
       "      <td>SRR10868967</td>\n",
       "      <td>C430</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>SAMN13813817</td>\n",
       "      <td>SRR10869319</td>\n",
       "      <td>C444</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>SAMN13813816</td>\n",
       "      <td>SRR10869320</td>\n",
       "      <td>C441</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>SAMN13813726</td>\n",
       "      <td>SRR10869321</td>\n",
       "      <td>C096</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>SAMN13813717</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>C004</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>SAMN13813716</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>C003</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "      <td>sputum</td>\n",
       "      <td>33.9249 S 18.4241 E</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        BioSample          Run  Name geo_loc_name_country  geo_loc_name  \\\n",
       "0    SAMN13813815  SRR10868963  C439         South Africa  South Africa   \n",
       "1    SAMN13813814  SRR10868964  C436         South Africa  South Africa   \n",
       "2    SAMN13813813  SRR10868965  C434         South Africa  South Africa   \n",
       "3    SAMN13813812  SRR10868966  C433         South Africa  South Africa   \n",
       "4    SAMN13813811  SRR10868967  C430         South Africa  South Africa   \n",
       "..            ...          ...   ...                  ...           ...   \n",
       "356  SAMN13813817  SRR10869319  C444         South Africa  South Africa   \n",
       "357  SAMN13813816  SRR10869320  C441         South Africa  South Africa   \n",
       "358  SAMN13813726  SRR10869321  C096         South Africa  South Africa   \n",
       "359  SAMN13813717  SRR10869322  C004         South Africa  South Africa   \n",
       "360  SAMN13813716  SRR10869323  C003         South Africa  South Africa   \n",
       "\n",
       "              Instrument  Platform           ReleaseDate  Collection_Date  \\\n",
       "0    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "1    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "2    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "3    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "4    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "..                   ...       ...                   ...              ...   \n",
       "356  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "357  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "358  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "359  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "360  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019   \n",
       "\n",
       "    isolation_source              lat_lon                    Organism  \n",
       "0             sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "1             sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "2             sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "3             sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "4             sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "..               ...                  ...                         ...  \n",
       "356           sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "357           sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "358           sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "359           sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "360           sputum  33.9249 S 18.4241 E  Mycobacterium tuberculosis  \n",
       "\n",
       "[361 rows x 12 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table_keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "525b1793-f74c-4e78-967e-ff9dd1a5162e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>geo_loc_name_country</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Collection_Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13813815</td>\n",
       "      <td>SRR10868963</td>\n",
       "      <td>C439</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN13813814</td>\n",
       "      <td>SRR10868964</td>\n",
       "      <td>C436</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN13813813</td>\n",
       "      <td>SRR10868965</td>\n",
       "      <td>C434</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN13813812</td>\n",
       "      <td>SRR10868966</td>\n",
       "      <td>C433</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN13813811</td>\n",
       "      <td>SRR10868967</td>\n",
       "      <td>C430</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>SAMN13813817</td>\n",
       "      <td>SRR10869319</td>\n",
       "      <td>C444</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>SAMN13813816</td>\n",
       "      <td>SRR10869320</td>\n",
       "      <td>C441</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>SAMN13813726</td>\n",
       "      <td>SRR10869321</td>\n",
       "      <td>C096</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>SAMN13813717</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>C004</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>SAMN13813716</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>C003</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        BioSample          Run Sample Name geo_loc_name_country  geo_loc_name  \\\n",
       "0    SAMN13813815  SRR10868963        C439         South Africa  South Africa   \n",
       "1    SAMN13813814  SRR10868964        C436         South Africa  South Africa   \n",
       "2    SAMN13813813  SRR10868965        C434         South Africa  South Africa   \n",
       "3    SAMN13813812  SRR10868966        C433         South Africa  South Africa   \n",
       "4    SAMN13813811  SRR10868967        C430         South Africa  South Africa   \n",
       "..            ...          ...         ...                  ...           ...   \n",
       "356  SAMN13813817  SRR10869319        C444         South Africa  South Africa   \n",
       "357  SAMN13813816  SRR10869320        C441         South Africa  South Africa   \n",
       "358  SAMN13813726  SRR10869321        C096         South Africa  South Africa   \n",
       "359  SAMN13813717  SRR10869322        C004         South Africa  South Africa   \n",
       "360  SAMN13813716  SRR10869323        C003         South Africa  South Africa   \n",
       "\n",
       "              Instrument  Platform           ReleaseDate  Collection_Date  \n",
       "0    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "1    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "2    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "3    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "4    Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "..                   ...       ...                   ...              ...  \n",
       "356  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "357  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "358  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "359  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "360  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "\n",
       "[361 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table[['BioSample', 'Run', 'Sample Name', 'geo_loc_name_country', 'geo_loc_name', 'Instrument', 'Platform', 'ReleaseDate', 'Collection_Date']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1df1a723-4512-4c21-97b5-6e4eb6f9b004",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a55ea83-4422-4aae-abac-0227c2b43d43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84dbf197-48c5-4ec5-9b85-5546d3833098",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "10e8982c-aad3-42ec-a24a-e3ccbeeac557",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ROLLINGDB_ID</th>\n",
       "      <th>BIOSAMPLE_ACCESSION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>RUN</th>\n",
       "      <th>DB_OF_ORIGIN</th>\n",
       "      <th>ISOLATION_DATE</th>\n",
       "      <th>ISOLATION_COUNTRY</th>\n",
       "      <th>ISOLATION_REGION</th>\n",
       "      <th>SPUTUM/CULTURE</th>\n",
       "      <th>UNPAIRED/PAIRED</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SPECIES</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>Combined_Runs</th>\n",
       "      <th>Num_Runs</th>\n",
       "      <th>OtherNames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>C001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>C002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>C003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>C004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>C005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ROLLINGDB_ID BIOSAMPLE_ACCESSION SAMPLE  RUN DB_OF_ORIGIN ISOLATION_DATE  \\\n",
       "698         C001                 NaN    NaN  NaN         CASS            NaN   \n",
       "699         C002                 NaN    NaN  NaN         CASS            NaN   \n",
       "700         C003                 NaN    NaN  NaN         CASS            NaN   \n",
       "701         C004                 NaN    NaN  NaN         CASS            NaN   \n",
       "702         C005                 NaN    NaN  NaN         CASS            NaN   \n",
       "\n",
       "    ISOLATION_COUNTRY ISOLATION_REGION  SPUTUM/CULTURE  UNPAIRED/PAIRED  \\\n",
       "698               NaN              NaN               1                1   \n",
       "699               NaN              NaN               1                1   \n",
       "700               NaN              NaN               1                1   \n",
       "701               NaN              NaN               1                1   \n",
       "702               NaN              NaN               1                1   \n",
       "\n",
       "    Platform Model                     SPECIES  TaxID ReleaseDate LoadDate  \\\n",
       "698      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "699      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "700      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "701      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "702      NaN   NaN  Mycobacterium tuberculosis    NaN         NaN      NaN   \n",
       "\n",
       "    Combined_Runs  Num_Runs OtherNames  \n",
       "698           NaN       NaN        NaN  \n",
       "699           NaN       NaN        NaN  \n",
       "700           NaN       NaN        NaN  \n",
       "701           NaN       NaN        NaN  \n",
       "702           NaN       NaN        NaN  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_metadata.query(\"DB_OF_ORIGIN=='CASS'\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7dbfd12-48d4-4e86-bff3-dedafa8602af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c8c07445-ff6e-406a-a870-4c43b61075f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "313"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_fastqs = glob.glob(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/C*\")\n",
    "len(cass_fastqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8c0f3bc-bb4b-4a30-8443-14641c183af0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(313, 54)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_fastqs_isolates = [os.path.basename(fName) for fName in cass_fastqs]\n",
    "problem_cass_fastqs = [os.path.basename(fName) for fName in glob.glob(f\"/n/data1/hms/dbmi/farhat/rollingDB/problem_fastqs/C*\")]\n",
    "\n",
    "len(cass_fastqs_isolates), len(problem_cass_fastqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5b34b6ea-dc8e-4546-816f-3b58a088d88a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54, 9)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table.loc[cass_mapping_table['Sample Name'].isin(problem_cass_fastqs)][['BioSample', 'Run', 'Sample Name', 'geo_loc_name_country', 'geo_loc_name', 'Instrument', 'Platform', 'ReleaseDate', 'Collection_Date']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8faee5ec-9fc2-4d1e-9550-be11dafffa91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(307, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>geo_loc_name_country</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Collection_Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13813810</td>\n",
       "      <td>SRR10868968</td>\n",
       "      <td>C429</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN13813725</td>\n",
       "      <td>SRR10868973</td>\n",
       "      <td>C094</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN13813804</td>\n",
       "      <td>SRR10868975</td>\n",
       "      <td>C412</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN13813803</td>\n",
       "      <td>SRR10868976</td>\n",
       "      <td>C408</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN13813802</td>\n",
       "      <td>SRR10868977</td>\n",
       "      <td>C407</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>South Africa</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2020-04-28T00:00:00Z</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      BioSample          Run Sample Name geo_loc_name_country  geo_loc_name  \\\n",
       "0  SAMN13813810  SRR10868968        C429         South Africa  South Africa   \n",
       "1  SAMN13813725  SRR10868973        C094         South Africa  South Africa   \n",
       "2  SAMN13813804  SRR10868975        C412         South Africa  South Africa   \n",
       "3  SAMN13813803  SRR10868976        C408         South Africa  South Africa   \n",
       "4  SAMN13813802  SRR10868977        C407         South Africa  South Africa   \n",
       "\n",
       "            Instrument  Platform           ReleaseDate  Collection_Date  \n",
       "0  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "1  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "2  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "3  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  \n",
       "4  Illumina HiSeq 2500  ILLUMINA  2020-04-28T00:00:00Z             2019  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_fastas = cass_mapping_table.loc[cass_mapping_table['Sample Name'].isin(cass_fastqs_isolates)][['BioSample', 'Run', 'Sample Name', 'geo_loc_name_country', 'geo_loc_name', 'Instrument', 'Platform', 'ReleaseDate', 'Collection_Date']].reset_index(drop=True)\n",
    "\n",
    "print(check_fastas.shape)\n",
    "\n",
    "check_fastas.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "80113d4a-849c-4be1-bf6a-0068efc4f2df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_lines(file_path):\n",
    "    # Use subprocess to call wc -l to count the number of lines in a file\n",
    "    process = subprocess.Popen(['wc', '-l', file_path], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    stdout, stderr = process.communicate()\n",
    "\n",
    "    if process.returncode != 0:\n",
    "        print(f\"Error counting lines in {file_path}: {stderr.decode().strip()}\")\n",
    "        return None\n",
    "\n",
    "    # Extract and return the line count from the output\n",
    "    line_count = int(stdout.decode().strip().split()[0])\n",
    "    return line_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5b1732b8-f898-410c-9810-f0da43b57562",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "50\n",
      "100\n",
      "150\n",
      "200\n",
      "250\n",
      "300\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(304, 3)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "good_files = []\n",
    "bad_files = []\n",
    "\n",
    "for i, row in check_fastas.iterrows():\n",
    "\n",
    "    internal_id = row['Sample Name']\n",
    "    run_id = row['Run']\n",
    "\n",
    "    # if not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\") or not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\"):\n",
    "\n",
    "    #     subprocess.run(f\"gunzip -c /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_R1.fastq.gz > /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\", shell=True)\n",
    "\n",
    "    #     subprocess.run(f\"gunzip -c /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_R2.fastq.gz > /n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\", shell=True)\n",
    "\n",
    "    # if i % 25 == 0:\n",
    "    #     print(f\"Finished unzipping sample {i}\")\n",
    "        \n",
    "    F1_LC_internal = count_lines(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\")\n",
    "    F2_LC_internal = count_lines(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\")\n",
    "\n",
    "    if os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_1.fastq\") and os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_2.fastq\"):\n",
    "        \n",
    "        F1_LC_public = count_lines(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_1.fastq\")\n",
    "        F2_LC_public = count_lines(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_2.fastq\")\n",
    "\n",
    "        if F1_LC_internal == F2_LC_internal == F1_LC_public == F2_LC_public:\n",
    "            good_files.append(run_id)\n",
    "        else:\n",
    "            bad_files.append(run_id)\n",
    "\n",
    "    if i % 50 == 0:\n",
    "        print(i)\n",
    "\n",
    "len(good_files), len(bad_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4fdd0c29-c799-4a6f-87c5-22387912fe8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SRR10869030', 'SRR10869302', 'SRR10869310']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d141ed4e-7238-47bd-a1b0-f7ad53d4299b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1fbae39-def1-4603-ab0a-c3e1f34659e6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "240c9b32-0f77-4d0b-8fbd-8513596fb3f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_headers_FASTQ(fName):\n",
    "    subprocess.run(f\"awk 'NR % 2 == 0' {fName} > {fName.replace('_1', '_noID_1').replace('_2', '_noID_2')}\", shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e46566d3-427f-458a-8504-5909919a9712",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_files(file1, file2):\n",
    "    result = subprocess.run(['diff', file1, file2], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    if result.returncode == 0:\n",
    "        # print(\"Files are the same.\")\n",
    "        return 0\n",
    "    else:\n",
    "        # print(\"Files are different.\")\n",
    "        return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "89034051-2d63-4ddb-b92c-b78f4384902a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_files('/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/C001/C001_noID_1.fastq', '/n/scratch/users/s/sak0914/smk_output/C001/SRR10869296/SRR10869296_noID_1.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "03742e6f-f7f2-4baa-84b3-28870331cfaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_files('/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/C429/C429_noID_1.fastq', '/n/scratch/users/s/sak0914/smk_output/C429/SRR10868968/SRR10868968_noID_1.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2c41fdbc-ad1d-48d1-84df-e2badaf441e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>FQ1</th>\n",
       "      <th>FQ2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C429</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ID FQ1 FQ2\n",
       "0  C429   1   1"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_compare_FASTQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2734618b-0111-4adb-a89b-40ed03d37512",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df_compare_FASTQ = pd.DataFrame(columns=['ID', 'FQ1', 'FQ2'])\n",
    "\n",
    "for i, row in check_fastas.iterrows():\n",
    "\n",
    "    internal_id = row['Sample Name']\n",
    "    run_id = row['Run']\n",
    "\n",
    "    if not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_1.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_1.fastq\")\n",
    "\n",
    "    if not os.path.isfile(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_2.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_2.fastq\")\n",
    "        \n",
    "    if not os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_1.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_1.fastq\")\n",
    "\n",
    "    if not os.path.isfile(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_2.fastq\"):\n",
    "        remove_headers_FASTQ(f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_2.fastq\")\n",
    "\n",
    "    cmp_1 = compare_files(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_1.fastq\", f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_1.fastq\")\n",
    "\n",
    "    cmp_2 = compare_files(f\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/{internal_id}/{internal_id}_noID_2.fastq\", f\"/n/scratch/users/s/sak0914/smk_output/{internal_id}/{run_id}/{run_id}_noID_2.fastq\")\n",
    "    \n",
    "    df_compare_FASTQ.loc[i, :] = [internal_id, cmp_1, cmp_2]\n",
    "    \n",
    "    if i % 50 == 0:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "314633eb-bf55-4b72-a52f-b7a5fac7b441",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for fName in glob.glob(\"/n/data1/hms/dbmi/farhat/rollingDB/fastq_db/*/*_noID*\"):\n",
    "#     os.remove(fName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "efe1ed8b-db5a-403c-b0ed-ffd07b3fe692",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for fName in glob.glob(\"/n/scratch/users/s/sak0914/smk_output/*/*/*_noID*\"):\n",
    "#     os.remove(fName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f0fd38-8512-4f6b-aa76-38c0868509e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bb84d1e3-91d0-4d33-b50e-ec90c661b653",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Run</th>\n",
       "      <th>FASTQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN13813815</td>\n",
       "      <td>SRR10868963</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN13813814</td>\n",
       "      <td>SRR10868964</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN13813813</td>\n",
       "      <td>SRR10868965</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN13813812</td>\n",
       "      <td>SRR10868966</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN13813811</td>\n",
       "      <td>SRR10868967</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>SAMN13813817</td>\n",
       "      <td>SRR10869319</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>SAMN13813816</td>\n",
       "      <td>SRR10869320</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>SAMN13813726</td>\n",
       "      <td>SRR10869321</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>SAMN13813717</td>\n",
       "      <td>SRR10869322</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>SAMN13813716</td>\n",
       "      <td>SRR10869323</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        BioSample          Run  FASTQ\n",
       "0    SAMN13813815  SRR10868963      1\n",
       "1    SAMN13813814  SRR10868964      1\n",
       "2    SAMN13813813  SRR10868965      1\n",
       "3    SAMN13813812  SRR10868966      1\n",
       "4    SAMN13813811  SRR10868967      1\n",
       "..            ...          ...    ...\n",
       "356  SAMN13813817  SRR10869319      1\n",
       "357  SAMN13813816  SRR10869320      1\n",
       "358  SAMN13813726  SRR10869321      1\n",
       "359  SAMN13813717  SRR10869322      1\n",
       "360  SAMN13813716  SRR10869323      1\n",
       "\n",
       "[361 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cass_mapping_table_for_megapipe = cass_mapping_table[['BioSample', 'Run']]\n",
    "cass_mapping_table_for_megapipe['FASTQ'] = 1\n",
    "cass_mapping_table_for_megapipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f5dc0d-6062-4ed4-8739-cb863118cbfb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "125e570b-1a62-4b5e-9011-4eca950a2f55",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_fasta_sequences(file1, file2):\n",
    "    # Read sequences from both files into sets\n",
    "    sequences1 = {str(record.seq) for record in SeqIO.parse(file1, \"fasta\")}\n",
    "    sequences2 = {str(record.seq) for record in SeqIO.parse(file2, \"fasta\")}\n",
    "    \n",
    "    # Compare the sets of sequences\n",
    "    if sequences1 == sequences2:\n",
    "        return 0\n",
    "        \n",
    "    return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f90c2cd7-adde-405c-b7b2-a36796411df0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kraken_db/PGAP/S0085-01.HybridAsm.PGAP.Genome.fasta kraken_db/Bakta/S0085-01.HybridAsm.Bakta.Genome.fasta\n",
      "kraken_db/PGAP/S0107-01.HybridAsm.PGAP.Genome.fasta kraken_db/Bakta/S0107-01.HybridAsm.Bakta.Genome.fasta\n"
     ]
    }
   ],
   "source": [
    "diff_fastas = {}\n",
    "alns_lst = []\n",
    "\n",
    "for fName in glob.glob(\"kraken_db/PGAP/*\"):\n",
    "    sample_ID = os.path.basename(fName).split('.')[0]\n",
    "\n",
    "    fName_2 = fName.replace('PGAP', 'Bakta')\n",
    "    assert os.path.isfile(fName_2)\n",
    "\n",
    "    return_val = compare_fasta_sequences(fName, fName_2)\n",
    "\n",
    "    if return_val == 1:\n",
    "        print(fName, fName_2)\n",
    "    \n",
    "    diff_fastas[sample_ID] = return_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "061dbf8d-2544-4fea-807b-d7ee803d3c21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(151, 151)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(os.listdir(\"kraken_db/PGAP\")), len(os.listdir(\"kraken_db/Bakta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c0ca6185-283b-48a0-a364-9bd6d6c6b993",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(151, 2)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_fastas_df = pd.DataFrame(diff_fastas, index=[0]).T.reset_index()\n",
    "diff_fastas_df.columns = ['sample', 'seq_diff']\n",
    "diff_fastas_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e79fe258-1e43-4961-998b-97fb97eea81b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>seq_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>S0085-01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>S0107-01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample  seq_diff\n",
       "85  S0085-01         1\n",
       "88  S0107-01         1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diff_fastas_df.query(\"seq_diff==1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4b3bd043-1e40-471c-a5a4-2ea8d29b6744",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_id = 'S0085-01'\n",
    "seq_PGAP = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/PGAP/{sample_id}.HybridAsm.PGAP.Genome.fasta\", \"fasta\")]\n",
    "seq_Bakta = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/Bakta/{sample_id}.HybridAsm.Bakta.Genome.fasta\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7d5b80a8-5e45-45a4-b770-e1d48334c9f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4417316, 4417384)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seq_PGAP[0][1]), len(seq_Bakta[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7a7b243d-1d81-4a6f-8d84-387036a6a206",
   "metadata": {},
   "outputs": [],
   "source": [
    "S0085_check = \"/n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/TRUST_PB_Set1/S0085-01.LR.Asm.fasta\"\n",
    "S0085_check_fastq = [(seq.id, seq.seq) for seq in SeqIO.parse(S0085_check, \"fasta\")]\n",
    "\n",
    "S0107_check = \"/n/data1/hms/dbmi/farhat/mm774/Projects/231121.MtbSetV3.151CI.CompleteAndSR.Asms/TRUST_PB_Set1/S0107-01.LR.Asm.fasta\"\n",
    "S0107_check_fastq = [(seq.id, seq.seq) for seq in SeqIO.parse(S0107_check, \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7bdd830f-457a-4751-b3c1-241eaf9afd68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4417384, 4416462)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(S0085_check_fastq[0][1]), len(S0107_check_fastq[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2e43e952-7228-4e0c-afb9-5f2e1a0c55fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_id = 'S0107-01'\n",
    "seq_PGAP = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/PGAP/{sample_id}.HybridAsm.PGAP.Genome.fasta\", \"fasta\")]\n",
    "seq_Bakta = [(seq.id, seq.seq) for seq in SeqIO.parse(f\"kraken_db/Bakta/{sample_id}.HybridAsm.Bakta.Genome.fasta\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ed996fc0-5235-4b58-99b0-80fefe78ca3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4416405, 4416462)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seq_PGAP[0][1]), len(seq_Bakta[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2496d0a6-c2e6-4fbe-b2cf-2c73ad281a18",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "13698c04-ae40-461e-bf0e-c8442835b903",
   "metadata": {},
   "source": [
    "# Create BED file to exclude sites with low EBR (refined low confidence) or are within 50 bp of a PE/PPE gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1138b10-acd0-43ad-be9e-2e151f750d40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude_fName_Marin = 'RLC_Regions.Plus.LowPmapK50E4.H37Rv'\n",
    "\n",
    "# RLC_regions = pd.read_csv(f\"references/ref_genome/{exclude_fName_Marin}.bed\", sep='\\t', header=None)\n",
    "# RLC_regions.columns = ['CHR', 'BEG', 'END']\n",
    "# RLC_regions['CHR'] = 'Chromosome'\n",
    "# len(RLC_regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "bc89c459-1f0c-4129-aac6-e2534da7b32c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19924\n"
     ]
    }
   ],
   "source": [
    "# exclude PE/PPE/PGRS genes, sites with EBR < 0.99, and homoplasic sites from Vargas et al., PNAS 2023 \n",
    "EBR_fName_Marin = 'EBR_Marin.bedgraph'\n",
    "\n",
    "df_EBR = pd.read_csv(f\"references/ref_genome/{EBR_fName_Marin}\", sep='\\t', header=None)\n",
    "df_EBR.columns = ['CHR', 'BEG', 'END', 'EBR']\n",
    "df_EBR['CHR'] = 'Chromosome'\n",
    "\n",
    "exclude_genes = H37Rv_regions.query(\"Name.str.startswith('PE') | Name.str.startswith('PPE') | Name.str.contains('PGRS')\").Name.values\n",
    "\n",
    "PE_PPE_BED_file_df = pd.DataFrame(columns=['CHR', 'BEG', 'END'])\n",
    "\n",
    "for i, gene in enumerate(exclude_genes):\n",
    "\n",
    "    # these are 1-indexed inclusive\n",
    "    gene_start, gene_end = H37Rv_regions.set_index('Name').loc[gene, ['Start', 'Stop']].values\n",
    "\n",
    "    # BED file is 0-indexed half open interval\n",
    "    # also extend the range by 50 base pairs on either side\n",
    "    gene_start = gene_start - 1 - 50\n",
    "    gene_end = gene_end + 50\n",
    "\n",
    "    PE_PPE_BED_file_df.loc[i, :] = ['Chromosome', gene_start, gene_end]\n",
    "\n",
    "homoplasy = pd.read_excel(\"./references/phylogeny/Vargas_PNAS_2023_homoplasy.xlsx\", sheet_name=None)\n",
    "\n",
    "if len(homoplasy.keys()) == 1:\n",
    "    homoplasy = homoplasy['Sheet1']\n",
    "\n",
    "homoplasic_sites_BED_file = pd.DataFrame(columns=['CHR', 'BEG', 'END'])\n",
    "\n",
    "for i, row in homoplasy.iterrows():\n",
    "    homoplasic_sites_BED_file.loc[i, :] = ['Chromosome', row['H37Rv Position']-1, row['H37Rv Position']]\n",
    "\n",
    "BED_file_df = pd.concat([PE_PPE_BED_file_df, df_EBR.query(\"EBR < 0.99\")[['CHR', 'BEG', 'END']], homoplasic_sites_BED_file]).sort_values([\"BEG\", \"END\"], ascending=True)\n",
    "print(len(BED_file_df))\n",
    "\n",
    "BED_file_df.sort_values(['BEG', 'END']).to_csv(\"./references/phylogeny/exclude_regions_pre_merge.bed\", sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "d4297029-fe33-4cf6-aeba-4ae3f66a96c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19924\n"
     ]
    }
   ],
   "source": [
    "bedtools merge -i /home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions_pre_merge.bed > /home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ca7c5d-a3af-45b8-ba87-4a321b00c4f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gunzip -c SAMEA3392614_full.vcf.gz | bcftools filter -e \"SVTYPE == 'INS' | SVTYPE == 'DEL' | IC > 0 | DC > 0\" | bedtools subtract -a '-' -b '/home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions_combined.bed' | SnpSift extractFields '-' POS REF ALT FILTER QUAL AF DP BQ MQ -e \"\" >  /n/scratch/users/s/sak0914/smk_output/SAMEA3392614/lineage/SNP_sites.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "29d0b62e-5348-4ed3-b7d0-379b688d5964",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>174</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>188</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>212</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>274</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3770</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4408093</td>\n",
       "      <td>4408094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3771</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4408919</td>\n",
       "      <td>4408920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3772</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4411370</td>\n",
       "      <td>4411400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3773</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4411469</td>\n",
       "      <td>4411471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3774</th>\n",
       "      <td>Chromosome</td>\n",
       "      <td>4411474</td>\n",
       "      <td>4411475</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3775 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1        2\n",
       "0     Chromosome        0       33\n",
       "1     Chromosome      174      185\n",
       "2     Chromosome      188      211\n",
       "3     Chromosome      212      244\n",
       "4     Chromosome      274      275\n",
       "...          ...      ...      ...\n",
       "3770  Chromosome  4408093  4408094\n",
       "3771  Chromosome  4408919  4408920\n",
       "3772  Chromosome  4411370  4411400\n",
       "3773  Chromosome  4411469  4411471\n",
       "3774  Chromosome  4411474  4411475\n",
       "\n",
       "[3775 rows x 3 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"/home/sak0914/Mtb_Megapipe/references/phylogeny/exclude_regions.bed\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca1abdd-e7f4-4461-ba55-df293f9c9b0d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e49414b-d8ea-4544-8993-77fb637e21d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "0498432b-ca33-455f-a4ea-9d0994294f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_SNP_matrix_from_VCF(isolate, qual_thresh=10, min_read_depth=5, BQ_thresh=20, MQ_thresh=30):\n",
    "\n",
    "    # for encoding ambiguous alleles as possibly both nucleotides\n",
    "    iupac_two_nuc_codes = {\n",
    "                            'AG': 'R',  # Purine (A or G)\n",
    "                            'CT': 'Y',  # Pyrimidine (C or T)\n",
    "                            'CG': 'S',  # Strong (C or G)\n",
    "                            'AT': 'W',  # Weak (A or T)\n",
    "                            'GT': 'K',  # Keto (G or T)\n",
    "                            'AC': 'M',  # Amino (A or C),\n",
    "                            'AA': 'A',  # the following are so that you don't get a key error, but they don't actually need to be merged\n",
    "                            'CC': 'C',\n",
    "                            'GG': 'G',\n",
    "                            'TT': 'T'\n",
    "                          }\n",
    "    \n",
    "    # keep only SNPs (non-length changing variants)\n",
    "    SNP_sites = pd.read_csv(f\"/n/scratch/users/s/sak0914/smk_output/{isolate}/lineage/SNP_sites.tsv.gz\", sep='\\t', compression='gzip').query(\"REF.str.len() == ALT.str.len()\").drop_duplicates().reset_index(drop=True)\n",
    "    \n",
    "    # these are REF alleles\n",
    "    SNP_sites.loc[SNP_sites['ALT']=='.', 'ALT'] = SNP_sites['REF']\n",
    "\n",
    "    # Reference\n",
    "    SNP_sites.loc[(SNP_sites['FILTER'].isin(['PASS', 'Amb'])) &\n",
    "                  (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                  (SNP_sites['DP'] >= min_read_depth) &\n",
    "                  (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                  (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                  (SNP_sites['AF'] <= 0.25), 'ALLELE'] = SNP_sites['REF']\n",
    "    \n",
    "    # high quality variants\n",
    "    SNP_sites.loc[(SNP_sites['FILTER']=='PASS') &\n",
    "                  (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                  (SNP_sites['DP'] >= min_read_depth) &\n",
    "                  (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                  (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                  (SNP_sites['AF'] > 0.75), 'ALLELE'] = SNP_sites['ALT']\n",
    "    \n",
    "    # ambiguous variants -- use the IUPAC codes representing the two nucleotides (REF and ALT) at a given site\n",
    "    # faster to not apply it to all rows, only on the ones you need\n",
    "    SNP_sites.loc[(SNP_sites['FILTER']=='Amb') &\n",
    "                  (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                  (SNP_sites['DP'] >= min_read_depth) &\n",
    "                  (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                  (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                  (SNP_sites['AF'] > 0.25), 'ALLELE'] = SNP_sites.loc[(SNP_sites['FILTER']=='Amb') &\n",
    "                                                                      (SNP_sites['QUAL'] >= qual_thresh) &\n",
    "                                                                      (SNP_sites['DP'] >= min_read_depth) &\n",
    "                                                                      (SNP_sites['BQ'] >= BQ_thresh) &\n",
    "                                                                      (SNP_sites['MQ'] >= MQ_thresh) &\n",
    "                                                                      (SNP_sites['AF'] > 0.25)].apply(lambda row: iupac_two_nuc_codes.get(''.join(sorted(row['REF'] + row['ALT'])), np.nan), axis=1)\n",
    "    \n",
    "    # fill the rest with N = no information\n",
    "    # this includes low coverage sites because we don't know what's happening there, so in the absence of other information, it could be anything\n",
    "    # also includes areas with deletions -- those shouldn't be REF, technically they are '-', which is treated the same as N\n",
    "    # gap characters ('-') and Ns are treated the same in iqtree\n",
    "    SNP_sites['ALLELE'] = SNP_sites['ALLELE'].fillna('N')\n",
    "    \n",
    "    SNP_sites['ISOLATE'] = isolate\n",
    "\n",
    "    # return a pivoted matrix\n",
    "    return SNP_sites.pivot(index='ISOLATE', columns='POS', values='ALLELE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "8dc73cdd-4282-4be0-b68d-ec7cfeafa763",
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_matrix_full = []\n",
    "\n",
    "for isolate in os.listdir(\"/n/scratch/users/s/sak0914/smk_output\"):\n",
    "\n",
    "    SNP_fName = f\"/n/scratch/users/s/sak0914/smk_output/{isolate}/lineage/SNP_sites.tsv.gz\"\n",
    "    \n",
    "    if os.path.isfile(SNP_fName):\n",
    "        SNP_matrix = create_SNP_matrix_from_VCF(isolate)\n",
    "        SNP_matrix_full.append(SNP_matrix)\n",
    "\n",
    "SNP_matrix_full = pd.concat(SNP_matrix_full)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "1859fc42-29da-4b9a-affb-c21d512cd448",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>POS</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>38</th>\n",
       "      <th>39</th>\n",
       "      <th>40</th>\n",
       "      <th>41</th>\n",
       "      <th>42</th>\n",
       "      <th>43</th>\n",
       "      <th>...</th>\n",
       "      <th>4411523</th>\n",
       "      <th>4411524</th>\n",
       "      <th>4411525</th>\n",
       "      <th>4411526</th>\n",
       "      <th>4411527</th>\n",
       "      <th>4411528</th>\n",
       "      <th>4411529</th>\n",
       "      <th>4411530</th>\n",
       "      <th>4411531</th>\n",
       "      <th>4411532</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISOLATE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>00-R1210</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3392614</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA8741574</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 3956879 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "POS          34      35      36      37      38      39      40      41       \\\n",
       "ISOLATE                                                                        \n",
       "00-R1210           G       T       G       T       G       G       A       A   \n",
       "SAMEA3392614       G       T       G       T       G       G       A       A   \n",
       "SAMEA8741574       G       T       G       T       G       G       A       A   \n",
       "\n",
       "POS          42      43       ... 4411523 4411524 4411525 4411526 4411527  \\\n",
       "ISOLATE                       ...                                           \n",
       "00-R1210           C       G  ...       A       G       A       T       A   \n",
       "SAMEA3392614       C       G  ...       A       G       A       T       A   \n",
       "SAMEA8741574       C       G  ...       A       G       A       T       A   \n",
       "\n",
       "POS          4411528 4411529 4411530 4411531 4411532  \n",
       "ISOLATE                                               \n",
       "00-R1210           C       G       T       C       G  \n",
       "SAMEA3392614       C       G       T       C       G  \n",
       "SAMEA8741574       C       G       T       C       G  \n",
       "\n",
       "[3 rows x 3956879 columns]"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_matrix_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "22030dac-d375-4427-99e4-0adddac40f78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3, 56550)"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove sites that are the same across all isolates because they have no signal\n",
    "SNP_matrix_no_constant = SNP_matrix_full.loc[:, SNP_matrix_full.nunique(dropna=False) > 1]\n",
    "SNP_matrix_no_constant.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "ed86286c-aae6-4d20-be4a-ec40c13b66a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>POS</th>\n",
       "      <th>1452</th>\n",
       "      <th>1849</th>\n",
       "      <th>4348</th>\n",
       "      <th>4349</th>\n",
       "      <th>4350</th>\n",
       "      <th>4351</th>\n",
       "      <th>4352</th>\n",
       "      <th>4353</th>\n",
       "      <th>4354</th>\n",
       "      <th>4355</th>\n",
       "      <th>...</th>\n",
       "      <th>4397736</th>\n",
       "      <th>4400246</th>\n",
       "      <th>4401732</th>\n",
       "      <th>4404346</th>\n",
       "      <th>4404960</th>\n",
       "      <th>4406116</th>\n",
       "      <th>4407200</th>\n",
       "      <th>4407588</th>\n",
       "      <th>4407927</th>\n",
       "      <th>4408923</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISOLATE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>00-R1210</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA3392614</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMEA8741574</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 56550 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "POS          1452    1849    4348    4349    4350    4351    4352    4353     \\\n",
       "ISOLATE                                                                        \n",
       "00-R1210           T       C       N       N       N       N       N       N   \n",
       "SAMEA3392614       G       A       G       A       G       G       A       T   \n",
       "SAMEA8741574       G       C       G       A       G       G       A       T   \n",
       "\n",
       "POS          4354    4355     ... 4397736 4400246 4401732 4404346 4404960  \\\n",
       "ISOLATE                       ...                                           \n",
       "00-R1210           N       N  ...       C       A       C       T       C   \n",
       "SAMEA3392614       A       T  ...       T       G       A       C       C   \n",
       "SAMEA8741574       A       T  ...       C       G       C       C       G   \n",
       "\n",
       "POS          4406116 4407200 4407588 4407927 4408923  \n",
       "ISOLATE                                               \n",
       "00-R1210           C       G       T       T       C  \n",
       "SAMEA3392614       C       A       C       G       T  \n",
       "SAMEA8741574       A       A       T       T       C  \n",
       "\n",
       "[3 rows x 56550 columns]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_matrix_no_constant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21083fd4-48a6-4cbe-9ee5-4a943f2e5426",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove sites where more than 10% of isolates have Ns -- this is a param that can be changed by the user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81bd624d-196c-451b-96b4-cd9a8ba4093f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "44d79928-8527-4154-8b69-807d8658aac0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'kraken_db/BAKTA/01_R1134.HybridAsm.Bakta.Genome.fasta'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(\"kraken_db/BAKTA/*\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f7874046-bbc7-4fe1-beb2-d30b61fae314",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'kraken_db/PGAP/01_R1134.HybridAsm.PGAP.Genome.fasta'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(\"kraken_db/PGAP/*\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7fa2b110-842a-4e6a-9314-26d7c07994cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'kraken_db/BAKTA/01_R1134.HybridAsm.BAKTA.Genome.fasta'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fName.replace('PGAP', 'BAKTA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c74ee7fa-c3e7-4092-8c7c-9b8c05f49c9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequences in the two FASTA files are the same.\n"
     ]
    }
   ],
   "source": [
    "# Example usage\n",
    "compare_fasta_sequences(\"kraken_db/01_R1134.HybridAsm.Bakta.Genome.fasta\", \"kraken_db/01_R1134.HybridAsm.PGAP.Genome.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3088982c-4713-4bda-aee8-622f88daa2f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4319, 1)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "todo_isolates = pd.read_csv(\"../data_cleaning/samples_for_F2.txt\", header=None, sep='\\t')\n",
    "todo_isolates.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3161ae9a-13e7-435e-b724-09354e2d9719",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3863</th>\n",
       "      <td>SAMN37457963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0\n",
       "3863  SAMN37457963"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "todo_isolates.loc[todo_isolates[0]=='SAMN37457963']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1884bc8e-a2e0-4eb6-9182-40b13bd36890",
   "metadata": {},
   "outputs": [],
   "source": [
    "python3 MtbQuantCNN/data_processing/variant_calling/F2_calculation.py data_cleaning/samples_for_F2.txt /n/data1/hms/dbmi/farhat/rollingDB/genomic_data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
